File: projects/7b2/workspace/GsProtein/parameter_finder/find_BG_parameters.m
% This script calculates the bond graph parameters for all reactions of the
% a given module. Specify the directory.
% based on SERCA model of Pan et al, which is based on Tran et al. (2009). 
% Parameters calculated in module's directory, by using the kinetic
% parameters and stoichiometric matrix.

% 19 Jul INITIAL CONDITIONS
% given dq/dt = A.q + c where A is a square matrix, use null(A) to find ICs
% of q. 
% A is constructed from linearising equations through lm, and
% contains BG parameters 
% K kappa are contained in c, but disregarding this
% as this is not dependent on q.

% return W from kinetic_parameters

clear;
clc;
close all;

%% booleans
write_parameters_file = true;
include_constraints = true;
include_type2_reactions = true;

%% Set directories
current_dir = pwd;
Idx_backslash = find(current_dir == filesep);
data_dir =  [current_dir filesep 'data' filesep];
output_dir = [current_dir filesep 'output' filesep];
modname = current_dir(Idx_backslash(end-1):Idx_backslash(end));
modname = modname(5:end-1);

if contains(current_dir, 'beta1') && false
    matstr = '_withR_LR_scheme4';
else
    matstr = '';
end

%% Define volumes
V_myo = 34.4; % pL
V = struct();
V.V_myo = V_myo;

%% Load forward matrix
if include_type2_reactions
    stoich_path = [data_dir sprintf('all_forward_matrix%s.txt',matstr)];
else
    stoich_path = [data_dir 'all_noType2_forward_matrix.txt'];
end
stoich_file_id = fopen(stoich_path,'r');

stoich_file_data = textscan(stoich_file_id,'%s','delimiter','\n');
fclose(stoich_file_id);

num_rows = length(stoich_file_data{1});
num_cols = sum(stoich_file_data{1}{1} == ',')+1;
dims = struct();
dims.num_cols = num_cols;
dims.num_rows = num_rows;

N_f = zeros(num_rows,num_cols);

for i_row = 1:num_rows
    line_str = stoich_file_data{1}{i_row};
    line_split = regexp(line_str,',','split');
    
    for i_col = 1:num_cols
        N_f(i_row,i_col) = str2double(line_split{i_col});
    end
end


%% Load reverse matrix
if include_type2_reactions
    stoich_path = [data_dir sprintf('all_reverse_matrix%s.txt',matstr)];
else
    stoich_path = [data_dir 'all_noType2_reverse_matrix.txt'];
end
stoich_file_id = fopen(stoich_path,'r');

stoich_file_data = textscan(stoich_file_id,'%s','delimiter','\n');
fclose(stoich_file_id);

num_rows = length(stoich_file_data{1}); % num of kappa + K
num_cols = sum(stoich_file_data{1}{1} == ',')+1; % num of reactions

N_r = zeros(num_rows,num_cols);

for i_row = 1:num_rows
    line_str = stoich_file_data{1}{i_row};
    line_split = regexp(line_str,',','split');
    
    for i_col = 1:num_cols
        N_r(i_row,i_col) = str2double(line_split{i_col});
    end
end

N_fT = transpose(N_f);
N_rT = transpose(N_r);

%% Calculate stoichiometric matrix
% I matrix to align with placement of kappa down the column.
% x-axis of stoich matrix (R1a, R1b etc) coincides with the kp km of that
% kinetic reaction
N = N_r - N_f;
N_T = N_rT - N_fT;

I = eye(num_cols);

M = [I N_fT; I N_rT];  

addpath(current_dir);
addpath(data_dir);
[k_kinetic, N_cT, K_C, W] = kinetic_parameters(M, include_type2_reactions, dims, V); %%%%%%%%%%%%%%%%
if ~include_constraints
    N_cT = [];
end

M = [M; N_cT];

% Calculate bond graph constants from kinetic parameters
% Note: units of kappa are fmol/s, units of K are fmol^-1
k = transpose([k_kinetic' K_C]);
lambdaW = exp(pinv(M) * log(k));

% Check that kinetic parameters are reproduced by bond graph parameters
k_sub = exp(M*log(lambdaW));
diff = (k - k_sub)./k;
error = sum(abs(diff));

% Check that there is a detailed balance constraint
Z = transpose(null(transpose(M),'r'));

N_rref = rref(N);
R_mat = null(N,'r');
kf = k_kinetic(1:num_cols);
kr = k_kinetic(num_cols+1:end);
K_eq = kf'./kr';
zero_est = R_mat'*(K_eq');
zero_est_log = R_mat'*log(K_eq');

if isempty(R_mat)
    warning('R_mat is empty: matrix is full rank');
end
%% Save bond graph parameters
lambda = lambdaW./W;
kappa = lambda(1:num_cols);
K = lambda(num_cols+1:end);

fID = fopen([data_dir 'rxnID.txt'], 'r');
rxnID = textscan(fID,'%s', 'delimiter','\n');
fclose(fID);
fID = fopen([data_dir 'Kname.txt'], 'r');
Kname = textscan(fID,'%s');
fclose(fID);

Knamed = struct();
for k = 1:length(Kname{1})
    % assign name to K values. store as struct for output in .mat.
    sname = Kname{1}{k};
    Knamed.(sname) = K(k);
end
kappa_named = struct();
for k = 1:length(rxnID{1})
    % assign name to K values. store as struct for output in .mat.
    sname = rxnID{1}{k};
    kappa_named.(['r' sname]) = kappa(k);
end



if write_parameters_file
    save([output_dir 'all_params.mat'],'kappa','K','k_kinetic', 'Knamed','kappa_named'); %
end

for j = 1:length(kappa)
    fprintf('var kappa_%s: fmol_per_sec {init: %g, pub: out};\n',rxnID{1}{j},kappa(j));
end
for j = 1:length(K)
    fprintf('var K_%s: per_fmol {init: %g, pub: out};\n',Kname{1}{j},K(j));
end

if true 
    if false
        for j=1:length(K)
            fprintf('var q_%s: fmol {init: 1e-13};\n',Kname{1}{j});
        end

        for j=1:length(kappa)
            fprintf('var v%s: fmol_per_sec;\n',rxnID{1}{j})
        end
        for j=1:length(K)
            fprintf('var mu_%s: J_per_mol;\n',Kname{1}{j});
        end
        for j=1:length(kappa)
            fprintf('v%s = kappa_%s*exp(mu_a/(R*T));\n',rxnID{1}{j}, rxnID{1}{j})
        end
        for j=1:length(K)
            fprintf('ode(q_%s, time) = p;\n',Kname{1}{j});
        end
    end
    disp(newline)
    disp('// Global value');
    for j=1:length(K)
        fprintf('var q_%s: fmol {pub: out};\n',Kname{1}{j});
    end
    disp('// From submodule');
    for j=1:length(K)
        fprintf('var q_%s_m%s: fmol {pub: in};\n',Kname{1}{j}, modname);
    end

    for j=1:length(K)
        fprintf('q_%s = q_%s_m%s + q_%s_init;\n',Kname{1}{j}, Kname{1}{j},...
            modname,Kname{1}{j});
    end
    
    disp(newline)
    disp('// Input from global environment');
    for j=1:length(K)
        fprintf('var q_%s_global: fmol {pub: in};\n',Kname{1}{j});
    end
    disp('// Output to global environment');
    for j=1:length(K)
        fprintf('var q_%s: fmol {init: 1e-16, pub: out};\n',Kname{1}{j});
    end
    
    for j=1:length(K)
        fprintf('mu_%s = R*T*ln(K_%s*q_%s_global);\n',Kname{1}{j},Kname{1}{j},Kname{1}{j});
    end
    
    disp(newline)
    fprintf('def map between environment and %s for\n',modname);
    fprintf('vars time and time;\n');
    for j=1:length(K)
        fprintf('vars q_%s_m%s and q_%s;\n',Kname{1}{j}, modname,Kname{1}{j});
        fprintf('vars q_%s and q_%s_global;\n',Kname{1}{j}, Kname{1}{j});
    end
    disp('enddef;');
end

disp(newline)

File: projects/7b2/workspace/cAMP/parameter_finder/find_BG_parameters.py
# This script calculates the bond graph parameters for all reactions of the
# a given module. Specify the directory.
# based on SERCA model of Pan et al, which is based on Tran et al. (2009). 
# Parameters calculated in module's directory, by using the kinetic
# parameters and stoichiometric matrix.

# 19 Jul INITIAL CONDITIONS
# given dq/dt = A.q + c where A is a square matrix, use null(A) to find ICs
# of q. 
# A is constructed from linearising equations through lm, and
# contains BG parameters 
# K kappa are contained in c, but disregarding this
# as this is not dependent on q.

# return W from kinetic_parameters

import os
import csv
import json
import math
import numpy as np
import sympy
from scipy.linalg import null_space
from kinetic_parameters_cAMP import kinetic_parameters


def read_IDs(path):
    data = []
    with open(path,'r') as f:
        reader = csv.reader(f)
        for row in reader:
            data.append(row[0])
        f.close()
    return data


def load_matrix(stoich_path):
    matrix = []
    with open(stoich_path,'r') as f:
        reader = csv.reader(f,delimiter=',')
        for row in reader:
            matrix.append([int(r) for r in row])
        f.close()
    return matrix


if __name__ == "__main__":

    ## booleans
    write_parameters_file = True
    include_constraints = True
    include_type2_reactions = True

    ## Set directories
    current_dir = os.getcwd()
    data_dir = current_dir + '\data'
    output_dir = current_dir + '\output'
    modname = os.path.dirname(current_dir).split('\\')[-1].split('_')[-1]

    if ('beta1' in current_dir) and False:
        matstr = '_withR_LR_scheme4'
    else:
        matstr = ''

    ## Define volumes
    V_myo = 34.4 # pL
    V = dict()
    V['V_myo'] = V_myo

    ## Load forward matrix
    if include_type2_reactions:
        stoich_path = data_dir + '\\all_forward_matrix%s.txt'%matstr
    else:
        stoich_path = data_dir + '\\all_noType2_forward_matrix.txt'

    N_f = load_matrix(stoich_path)


    ## Load reverse matrix
    if include_type2_reactions:
        stoich_path = data_dir + '\\all_reverse_matrix%s.txt'%matstr
    else:
        stoich_path = data_dir + '\\all_noType2_reverse_matrix.txt'
    
    N_r = load_matrix(stoich_path)

    N_fT = np.transpose(N_f)
    N_rT = np.transpose(N_r)

    ## Calculate stoichiometric matrix
    # I matrix to align with placement of kappa down the column.
    # x-axis of stoich matrix (R1a, R1b etc) coincides with the kp km of that kinetic reaction
    N = [[N_r[j][i] - N_f[j][i] for i in range(len(N_f[0]))] for j in range(len(N_f))]
    N_T = [[N_rT[j][i] - N_fT[j][i] for i in range(len(N_fT[0]))] for j in range(len(N_fT))]

    num_rows = len(N)
    num_cols = len(N[0])
    dims = dict()
    dims['num_rows'] = num_rows
    dims['num_cols'] = num_cols

    I = np.identity(num_cols)

    M = np.append(np.append(I, N_fT,1), np.append(I, N_rT,1),0)

    # addpath(current_dir)
    # addpath(data_dir)
    [k_kinetic, N_cT, K_C, W] = kinetic_parameters(M, include_type2_reactions, dims, V) ################
    if not include_constraints:
        N_cT = []

    try:
        M = np.append(M, N_cT,0)
        k = np.append(k_kinetic, K_C, 0)
    except:
        k = k_kinetic

    # Calculate bond graph constants from kinetic parameters
    # Note: units of kappa are fmol/s, units of K are fmol^-1
    lambda_expo = np.matmul(np.linalg.pinv(M), [math.log(ik) for ik in k])
    lambdaW = [math.exp(l) for l in lambda_expo]


    # Check that kinetic parameters are reproduced by bond graph parameters
    k_est = np.matmul(M,[math.log(k) for k in lambdaW])
    k_est = [math.exp(k) for k in k_est]
    diff = [(k[i] - k_est[i])/k[i] for i in range(len(k))]

    error = np.sum([abs(d) for d in diff])


    # Check that there is a detailed balance constraint
    Z = null_space(np.transpose(M))

    N_rref = sympy.Matrix(N).rref()
    R_mat = null_space(N)
    kf = k_kinetic[:num_cols]
    kr = k_kinetic[num_cols:]
    K_eq = [kf[i]/kr[i] for i in range(len(kr))]
    zero_est = np.matmul(np.transpose(R_mat),K_eq)
    zero_est_log = np.matmul(np.transpose(R_mat),[math.log(k) for k in K_eq])

    # if not R_mat:
    #     warning('R_mat is empty: matrix is full rank')
    
    lambdak = [lambdaW[i]/W[i] for i in range(len(W))]
    kappa = lambdak[:len(N[0])]
    K = lambdak[len(N[0]):]

    rxnID = read_IDs('data\\rxnID.txt')
    Kname = read_IDs('data\\Kname.txt')


    # ### print outputs ###
    for ik in range(len(kappa)):
        print('var kappa_%s: fmol_per_sec {init: %g, pub: out};' %(rxnID[ik],kappa[ik]))
    for ik in range(len(Kname)):
        print('var K_%s: per_fmol {init: %g, pub: out};' %(Kname[ik],K[ik]))

    file = open(output_dir + '/all_parameters_out.json', 'w')
    data = { "K": K, "kappa": kappa, "k_kinetic": k_kinetic }
    json.dump(data, file)


    if True:
        if False:
            for j in range(len(K)):
                print('var q_%s: fmol {init: 1e-13};'%Kname[j])
            for j in range(len(kappa)):
                print('var v%s: fmol_per_sec;'%rxnID[j])
            for j in range(len(K)):
                print('var mu_%s: J_per_mol;'%Kname[j])
            for j in range(len(kappa)):
                print('v%s = kappa_%s*exp(mu_a/(R*T));'%rxnID[j], rxnID[j])
            for j in range(len(K)):
                print('ode(q_%s, time) = p;'%Kname[j])
        print('\n')
        print('// Global value')
        for j in range(len(K)):
            print('var q_%s: fmol {pub: out};'%Kname[j])
        print('// From submodule')
        for j in range(len(K)):
            print('var q_%s_m%s: fmol {pub: in};'%(Kname[j], modname))

        for j in range(len(K)):
            print('q_%s = q_%s_m%s + q_%s_init;'%(Kname[j], Kname[j],
                modname,Kname[j]))
        print('\n')
        print('// Input from global environment')
        for j in range(len(K)):
            print('var q_%s_global: fmol {pub: in};'%Kname[j])
        print('// Output to global environment')
        for j in range(len(K)):
            print('var q_%s: fmol {init: 1e-16, pub: out};'%(Kname[j]))
        
        for j in range(len(K)):
            print('mu_%s = R*T*ln(K_%s*q_%s_global);'%(Kname[j],Kname[j],Kname[j]))
        print('\n')
        print('def map between environment and %s for'%modname)
        print('vars time and time;')
        for j in range(len(K)):
            print('vars q_%s_m%s and q_%s;'%(Kname[j], modname,Kname[j]))
            print('vars q_%s and q_%s_global;'%(Kname[j], Kname[j]))
        print('enddef;')



